core-rust/state-manager/src/staging/cache.rs

```diff
@@ -147,20 +166,74 @@ impl ExecutionCache {
             }
         };
     }
+
+    fn process<S: RootStore>(
+        store: &StagedStore<S>,
+        transaction_receipt: TransactionReceipt,
+    ) -> ProcessedResult {
+        let hash_changes = match &transaction_receipt.result {
+            TransactionResult::Commit(commit) => commit
+                .state_updates
+                .up_substates
+                .iter()
+                .map(|(id, value)| {
+                    (
+                        id.clone(),
+                        Some(hash(scrypto_encode(&value.substate).unwrap())),
+                    )
+                })
+                .collect::<Vec<(SubstateId, Option<Hash>)>>(),
+            TransactionResult::Reject(_) | TransactionResult::Abort(_) => Vec::new(),
+        };
+        let mut collector = CollectingTreeStore::new(store);
+        let state_version = store.overlay.state_version;
+        let root_hash = put_at_next_version(&mut collector, state_version, &hash_changes);
+        ProcessedResult::new(root_hash, transaction_receipt, collector.diff)
+    }
+}
+
+struct CollectingTreeStore<'s, S: ReadableTreeStore> {
+    readable_delegate: &'s S,
+    diff: HashTreeDiff,
+}
+
+impl<'s, S: ReadableTreeStore> CollectingTreeStore<'s, S> {
+    pub fn new(readable_delegate: &'s S) -> Self {
+        Self {
+            readable_delegate,
```

Comment:
Can we give this a better name than `readable_delegate`? Is this actually the root store?